### PR TITLE
feat: add assignee filter for GitHub sensor (issues, PRs, projects)

### LINKED ISSUE #2: Add assignee filter for github issues/projects/PRs
Enable assignee filtering for the github sensor for the issues, PRs and projects types.

### PR DESCRIPTION
**Pull request by adamveld12**
## Summary

Closes #2

Adds full assignee filtering parity between the Linear and GitHub sensors — issues, pull requests, and project items can all be filtered by GitHub login username via `filter_assignees` (or sensor-level `assignees`).

### Changes

- **`sensors/github/src/types.ts`** — Implements `matchesGitHubIssueFilters`, `matchesGitHubPRFilters`, and `matchesGitHubProjectItemFilters` with `filter_assignees` support. All three functions use case-insensitive login comparison (`.toLowerCase()`) for consistency.
- **`sensors/github/src/backend.ts`** — `fetchOne` for projects now skips the GraphQL assignee fetch (state refresh only; no filter re-evaluation needed).
- **`sensors/github/src/schema.ts`** — Adds `assignees` field to the GitHub sensor schema; propagated as default `filter_assignees` to workflows.
- **`docs/references/sensors.md`** — Documents the GitHub sensor `assignees` field; trims table column widths.

### Code Review Fixes (latest review on `65556295`)

- 🔴 **CRITICAL**: Removed duplicate/broken block in `matchesGitHubProjectItemFilters` that caused `TS1128: Declaration or statement expected`.
- 🟡 **MEDIUM**: Normalized all assignee comparisons to lowercase — `matchesGitHubIssueFilters` and `matchesGitHubPRFilters` now match the case-insensitive behavior of `matchesGitHubProjectItemFilters`.
- 🟡 **MEDIUM**: `fetchOne` in `createGitHubProjectsBackend` no longer triggers the GraphQL assignee fetch — passes `resolveAssignees=false` since state refresh doesn't require filter re-evaluation.
- 🟠 **TRIVIAL**: Trimmed overly wide description cells in the GitHub sensor docs table.

### CODE REVIEW

**Review by adamveld12 — COMMENTED** [review]
## Code Review Summary

CI ✅ — all checks pass.

### 🔴 CRITICAL
None.

### 🟡 MEDIUM

1. **Unnecessary GraphQL call on every poll** — `fetchProjectItems` unconditionally calls `fetchProjectItemAssignees` even when no `filter_assignees` is configured. This doubles the API calls per poll cycle for every projects-mode workflow, adding latency and burning rate limit budget for no benefit.

2. **`ghApiGraphQL` duplicates `spawnGh`** — The new function reimplements the env setup, `Bun.spawn`, stdout/stderr collection, and exit code check already handled by `spawnGh`. Should reuse the existing helper and add stdin support to it.

3. **Overly complex type annotation** — The conditional mapped type on the `items` variable inside `fetchProjectItemAssignees` is unreadable and provides no safety beyond `const items = data.repositoryOwner?.projectV2?.items;`. Replace with the simple optional chain.

4. **No pagination safety limit** — `fetchProjectItemAssignees` will loop indefinitely if the GitHub API keeps returning `hasNextPage: true` (e.g. a bug or unexpected behavior). Add a max-pages guard (e.g. 50) consistent with defensive patterns elsewhere.

### 🟠 TRIVIAL

1. **`assignees` optionality mismatch** — `GitHubProjectItemNode.assignees` is typed `optional`, but `fetchProjectItems` always sets it (to `assigneeMap.get(item.id) ?? []`). Consider making it required to match runtime behavior and avoid unnecessary `?? []` / `undefined` checks downstream.

**Review by adamveld12 — COMMENTED** [review]
## Code Review Summary

### 🔴 CRITICAL

None. CI passes, no security issues or show-stoppers found.

### 🟡 MEDIUM

1. **Workflow-level `filter_assignees` without sensor `assignees` silently drops all project items** — `resolveAssignees` is computed once at sensor creation in `factory.ts` based on `config.assignees?.length`. If a workflow sets `filter_assignees` in its tracker config but the sensor has no `assignees`, `resolveAssignees` will be `false`, the GraphQL call is skipped, all project items get empty `assignees` arrays, and `matchesGitHubProjectItemFilters` rejects everything. This is a silent data loss bug.

2. **Case-sensitive login comparison** — GitHub logins are case-insensitive (`Alice` == `alice`), but `matchesGitHubProjectItemFilters` (and the existing issue/PR filters) use `assigneeLogins.includes(fa)` which is case-sensitive. A user configuring `"Alice"` when the API returns `"alice"` will get zero matches.

3. **GraphQL `assignees(first: 10)` silently truncates** — Items with >10 assignees will lose data. This is unlikely in practice but the silent truncation could cause hard-to-debug filter misses on large teams. Consider either increasing the limit or logging when `hasNextPage` is true on the assignees connection.

### 🟠 TRIVIAL

1. **Duplicate assignee filter logic** — The assignee check in `matchesGitHubIssueFilters`, `matchesGitHubPRFilters`, and `matchesGitHubProjectItemFilters` is identical. Could be extracted into a shared `matchesAssigneeFilter(logins: string[], config: TrackerConfig)` helper.

2. **Docs table formatting** — The `assignees` row description in `sensors.md` is very long and creates a wide table. Consider shortening the inline description and linking to the guide for details.

**Comment by adamveld12** [comment]
## Code Review — `feat: add assignee filter for GitHub sensor`

### 🔴 CRITICAL (1)

1. **Duplicate/broken code in `matchesGitHubProjectItemFilters` — CI failing**
   `sensors/github/src/types.ts` has duplicate assignee-check logic with a stray closing brace, causing `TS1128: Declaration or statement expected`. The function has the correct implementation inside the `if` block (with `.toLowerCase()`), then a second dangling copy outside it (without `.toLowerCase()`) that creates a syntax error. This is the sole cause of the CI failure.

   **Fix:** Remove lines 206–208 (the duplicate block outside the `if`):
   ```diff
   export function matchesGitHubProjectItemFilters(node: GitHubProjectItemNode, config: TrackerConfig): boolean {
     if (config.filter_assignees?.length) {
       const assigneeLogins = node.assignees.map((a) => a.login.toLowerCase());
       if (!config.filter_assignees.some((fa) => assigneeLogins.includes(fa.toLowerCase()))) return false;
     }
   -    const assigneeLogins = node.assignees.map((a) => a.login);
   -    if (!config.filter_assignees.some((fa) => assigneeLogins.includes(fa))) return false;
   -  }
     return true;
   }
   ```

---

### 🟡 MEDIUM (2)

1. **Case-sensitivity inconsistency across filter functions**
   `matchesGitHubProjectItemFilters` (the correct block) uses `.toLowerCase()` for case-insensitive matching, but `matchesGitHubIssueFilters` and `matchesGitHubPRFilters` do exact-case matching. GitHub logins are case-insensitive, so all three should behave the same way. Pick one strategy and apply it consistently.

2. **`fetchOne` re-fetches all items + GraphQL assignees**
   `backend.ts` `fetchOne()` calls `fetchProjectItems(owner, projectNumber, token, resolveAssignees)` which fetches ALL project items and optionally runs the full GraphQL pagination just to find a single item by ID. For large projects this is wasteful on every single-item lookup. Consider caching the last poll result, or passing `resolveAssignees: false` in `fetchOne` if it's only used for state-change checks.

---

### 🟠 TRIVIAL (1)

1. **Docs table column width explosion** — The `sensors.md` reference table now has an extremely wide Description column for the `assignees` row. Consider shortening the inline description and linking to the guide for details.

**Review by adamveld12 — COMMENTED** [review]
See review summary in PR comments.

**Review by adamveld12 — COMMENTED** [review]
## Code Review: feat: add assignee filter for GitHub sensor

CI ✅ — all checks passing.

---

### 🟡 MEDIUM

**1. Silent filter failure for project-mode workflows with per-workflow `filter_assignees`**

`resolveAssignees` in `factory.ts` is determined at sensor creation time based solely on `config.assignees?.length`. If a workflow sets `tracker.filter_assignees` but the sensor has no `assignees`, the GraphQL assignee fetch is skipped → all project items get empty `assignees` arrays → the filter silently rejects everything.

This doesn't affect `issues` or `pull_requests` modes (REST API always returns assignees), only `projects` mode where assignee data requires the extra GraphQL call.

**Suggested fix:** Either document this limitation explicitly, or defer the `resolveAssignees` decision to the backend's `fetchAll()` by passing the tracker config (or a flag) at fetch time rather than at sensor construction.

**2. No case-insensitive test coverage**

`.toLowerCase()` was added to all three filter functions — good hardening — but no test verifies mixed-case matching (e.g. `"Alice"` in the filter matching `"alice"` on the node). Adding one case per function would lock in the behavior.

**3. Type-unsafe config access in `sensor-manager.ts`**

`rawConfig as Record<string, unknown>` + `rc["assignees"] as string[]` bypasses type safety. This is consistent with the existing Linear propagation pattern directly above it, but both would benefit from a small helper (e.g. `getSensorField<T>(rawConfig, field)`) to reduce duplication and improve safety. Not blocking — just a quality-of-life note for a future pass.

---

### 🟠 TRIVIAL

**4. Duplicate test in `matchesGitHubProjectItemFilters`**

`"returns false when filter is set but node has empty assignees"` and `"returns false when node has empty assignees array but filter is set"` are identical — same setup, same assertion. One can be removed.

---

### Summary

Solid implementation overall. The GraphQL pagination with `MAX_PAGES` guard, graceful error handling in `fetchProjectItemAssignees`, and the `fetchOne` optimization (skipping the GraphQL call for state refreshes) are all well done. Docs are thorough and consistent across all three doc tiers.

The only finding with functional risk is **#1** (project-mode edge case). The rest are test/quality polish.

File: sensors/github/src/factory.ts [hunk 1]
```diff
@@ -13,7 +13,8 @@ export function createGitHubSensor(config: GitHubSensorConfig): Sensor {
         `GitHub sensor (${owner}/${repo}) in "projects" mode requires a "project" name in the sensor config.`,
       );
     }
-    const backend = createGitHubProjectsBackend(owner, config.project, token);
+    const resolveAssignees = (config.assignees?.length ?? 0) > 0;
```
Comment (adamveld12) [inline sensors/github/src/factory.ts]: 🟡 **MEDIUM**: `resolveAssignees` is determined here at sensor creation time based solely on `config.assignees?.length`. If a workflow sets `tracker.filter_assignees` without the sensor having `assignees`, this will be `false` → the GraphQL assignee fetch is skipped → all project items get empty assignee arrays → filter silently rejects everything.

Consider deferring this decision to `fetchAll()` time, or documenting that project-mode assignee filtering requires the sensor-level `assignees` field to be set.

File: sensors/github/src/__tests__/types.test.ts [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+  test("returns true when assignee login matches one of filter_assignees (OR logic)", () => {
+    const issue = makeIssue({ assignees: [{ login: "alice" }] });
+    const config = baseTrackerConfig({ filter_assignees: ["alice", "bob"] });
+    expect(matchesGitHubIssueFilters(issue, config)).toBe(true);
+  });
+
+  test("returns false when no assignee matches filter_assignees", () => {
+    const issue = makeIssue({ assignees: [{ login: "charlie" }] });
+    const config = baseTrackerConfig({ filter_assignees: ["alice", "bob"] });
+    expect(matchesGitHubIssueFilters(issue, config)).toBe(false);
```
Comment (adamveld12) [inline sensors/github/src/__tests__/types.test.ts]: 🟡 **MEDIUM**: Missing case-insensitive test coverage. The code adds `.toLowerCase()` to all three filter functions but no test verifies mixed-case matching. Suggest adding one test per function, e.g.:

```suggestion
  test("matches case-insensitively", () => {
    const issue = makeIssue({ assignees: [{ login: "Alice" }] });
    const config = baseTrackerConfig({ filter_assignees: ["alice"] });
    expect(matchesGitHubIssueFilters(issue, config)).toBe(true);
  });
```

File: sensors/github/src/__tests__/types.test.ts [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+    const config = baseTrackerConfig({ filter_assignees: ["alice"] });
+    expect(matchesGitHubProjectItemFilters(item, config)).toBe(false);
+  });
+
+  test("returns true when multiple assignees and at least one matches", () => {
+    const item = makeProjectItem({ assignees: [{ login: "charlie" }, { login: "bob" }] });
+    const config = baseTrackerConfig({ filter_assignees: ["alice", "bob"] });
+    expect(matchesGitHubProjectItemFilters(item, config)).toBe(true);
+  });
+});
```
Comment (adamveld12) [inline sensors/github/src/__tests__/types.test.ts]: 🟠 **TRIVIAL**: `"returns false when filter is set but node has empty assignees"` (line 471 in diff) and `"returns false when node has empty assignees array but filter is set"` (line 478) are identical — same setup, same assertion. One can be removed.
